assets/js/app.ts

```diff
@@ -1,13 +1,85 @@
-// This is a manifest file that'll be compiled into including all the files listed below.
-// Add new JavaScript/Coffee code in separate files in this directory and they'll automatically
-// be included in the compiled file accessible from http://example.com/assets/application.js
-// It's not advisable to add code directly here, but if you do, it'll appear at the bottom of the
-// the compiled file.
+// This is the main application file. Its job is to import any modules that modify global state, and run
+// all other functions when the DOM is ready using the `whenReady` utility function.
 //
+// Import new files and add their exported setup function to the functions array below.
 
-// Our code
 import './ujs';
-import './when-ready';
+
+import { whenReady } from './utils/dom';
+
+import { listenAutocomplete } from './autocomplete';
+import { loadBooruData } from './booru';
+import { registerEvents } from './boorujs';
+import { setupBurgerMenu } from './burger';
+import { bindCaptchaLinks } from './captcha';
+import { setupComments } from './comment';
+import { setupDupeReports } from './duplicate_reports';
+import { setSesCookie } from './fp';
+import { setupGalleryEditing } from './galleries';
+import { initImagesClientside } from './imagesclientside';
+import { bindImageTarget } from './image_expansion';
+import { setupEvents } from './misc';
+import { setupNotifications } from './notifications';
+import { setupPreviews } from './preview';
+import { setupQuickTag } from './quick-tag';
+import { setupSettings } from './settings';
+import { listenForKeys } from './shortcuts';
+import { initTagDropdown } from './tags';
+import { setupTagListener } from './tagsinput';
+import { setupTagEvents } from './tagsmisc';
+import { setupTimestamps } from './timeago';
+import { setupImageUpload } from './upload';
+import { setupSearch } from './search';
+import { setupToolbar } from './markdowntoolbar';
+import { hideStaffTools } from './staffhider';
+import { pollOptionCreator } from './poll';
+import { warnAboutPMs } from './pmwarning';
+import { imageSourcesCreator } from './sources';
+
+const functions = [
+  loadBooruData,
+  listenAutocomplete,
+  registerEvents,
+  setupBurgerMenu,
+  bindCaptchaLinks,
+  initImagesClientside,
+  setupComments,
+  setupDupeReports,
+  setSesCookie,
+  setupGalleryEditing,
+  bindImageTarget,
+  setupEvents,
+  setupNotifications,
+  setupPreviews,
+  setupQuickTag,
+  setupSettings,
+  listenForKeys,
+  initTagDropdown,
+  setupTagListener,
+  setupTagEvents,
+  setupTimestamps,
+  setupImageUpload,
+  setupSearch,
+  setupToolbar,
+  hideStaffTools,
+  pollOptionCreator,
+  warnAboutPMs,
+  imageSourcesCreator,
+];
+
+whenReady(() => {
+  functions.forEach(fn => {
+    try {
+      fn();
+    } catch (err: unknown) {
+      console.log(`${fn.name} ran with errors.`);
+
+      if (err instanceof Error) {
+        console.log(`The error was:\n\n${err.message}`);
+      }
+    }
+  });
+});
```

Comment:
This logic should be moved into whenReady.

e.g., whenReady should be declared as
`export function whenReady(...callbacks: VoidFunction[]) { for (const callback of callbacks) { ...  } }`

Then app.ts can simply do:

```ts
whenReady(
  loadBooruData,
  listenAutocomplete,
  // ...
);
```

That way this module (app.ts) will end up having only one testable source line, so excluding it from coverage becomes more reasonable.